optimizer/lib/transformers/BlurImagePlaceholders.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const sizeOf = require('image-size');
+const jimp = require('jimp');
+
+/**
+ * Adds placeholders for all AMP images and posters that are blurry versions of
+ * the corresponding original source. The blur will be displayed as the
+ * <amp-img> is rendering, and will fade out once the element is loaded.
+ */
+
+class BlurImagePlaceholders {
+  /**
+   * Parses the document to add blurred placedholders in all appropriate
+   * locations.
+   * @param {Tree} tree - a DOM tree.
+   * @param {Object} params - a dictionary containing transformer specific parameters.
+   */
+  transform(tree, params) {
+    const html = tree.root.firstChildByTag('html');
+    const body = html.firstChildByTag('body');
+    const promises = [];
+    for (let node = body; node !== null; node = node.nextNode()) {
+      let childSrc;
+      if (node.tagName === 'amp-img') {
+        childSrc = node.attribs.src;
+      }
+      if (node.tagName === 'amp-video' && node.attribs.poster) {
+        childSrc = node.attribs.poster;
+      }
+      if (childSrc) {
+        let parent = node;
+        promises.push(this.addBitmap_(tree, childSrc).then(imgChild => {
+          parent.appendChild(imgChild);
+        }));
+      }
+    }
+    return Promise.all(promises);
+  }
+
+  /**
+   * Adds a child image
+   * @param {Tree} tree - a DOM tree
+   * @param {String} parentURL - The image that the bitmap is based on.
+   */
+  addBitmap_(tree, parentImg) {
```

Comment:
We're not checking for an existing placeholder yet.

Reply:
Updated the file to do the check! The code that does so is in the hasPlaceholder_() method at the end